src/NuGet.Core/NuGet.Protocol.Core.v3/PackageUpdateResource.cs

```diff
@@ -0,0 +1,601 @@
+        private static bool IsFileSource(string source)
+        {
+            Uri uri;
+            if (Uri.TryCreate(source, UriKind.RelativeOrAbsolute, out uri))
+            {
+                return uri.IsFile;
+            }
+            else
+            {
+                return false;
+            }
+        }
+
+        // Pushes a package to the Http server.
+        private async Task PushPackageToServer(string source,
+            string apiKey,
+            string pathToPackage,
+            long packageSzie,
+            ILogger logger,
+            CancellationToken token)
+        {
+            HttpRequestMessage request = null;
+            HttpResponseMessage response = null;
+
+            try
+            {
+                response = await _httpSource.SendAsync(
+                    () => { return request = CreateRequest(request, source, pathToPackage, apiKey); },
+                    token);
+            }
+            finally
+            {
+                if (request != null)
+                {
+                    request.Dispose();
+                }
+            };
+
+            response.EnsureSuccessStatusCode();
+        }
+
+        private HttpRequestMessage CreateRequest(HttpRequestMessage currentRequest,
+            string source,
+            string pathToPackage,
+            string apiKey)
+        {
+            if (currentRequest != null)
+            {
+                //this should dispose the content, the file stream underneath, and everything.
+                currentRequest.Dispose();
+            }
+
+            var fileStream = new FileStream(pathToPackage, FileMode.Open, FileAccess.Read, FileShare.Read);
+            var request = new HttpRequestMessage(HttpMethod.Put, GetServiceEndpointUrl(source, string.Empty));
+            var content = new MultipartFormDataContent();
+
+            var packageContent = new StreamContent(fileStream);
+            packageContent.Headers.ContentType = MediaTypeHeaderValue.Parse("application/octet-stream");
+            content.Add(packageContent, "package", "package.nupkg");
+            request.Content = content;
+
+            if (!string.IsNullOrEmpty(apiKey))
+            {
+                request.Headers.Add(ApiKeyHeader, apiKey);
+            }
+            return request;
+        }
+
+        private void PushPackageToFileSystem(Uri sourceUri, string pathToPackage)
+        {
+            string root = sourceUri.LocalPath;
+            PackageArchiveReader reader = new PackageArchiveReader(pathToPackage);
+            PackageIdentity packageIdentity = reader.GetIdentity();
+
+            //TODD: support V3 repo style if detect it is
+            var pathResolver = new PackagePathResolver(root, useSideBySidePaths: true);
+            var packageFileName = pathResolver.GetPackageFileName(packageIdentity);
+
+            string fullPath = Path.Combine(root, packageFileName);
```

Comment:
it is needed, the "packageFileName" is just a file name, and we will copy to the repo root, which  is the correct v2 behavior. This method will be changed in the next PR that to incorporate the v3 behavior using the same support by AddCommand.